src/modules/jalousie/geometry/DreamGeometry.h

```diff
@@ -70,52 +239,106 @@ uint32_t PixelIdFromSlicePixel(SlicePixel Slice) {
     { 48,  8, 3 }, { 48,  8, 3 },                                            // 12-13, SlicePixel.X 48-55: Sumo 3, Cols 1-8
   };
   // clang-format on
-  uint32_t XCompact =
-      Slice.X / 4; // Range reduced from 56 to 14 -> fewer constants
+  uint32_t XCompact = Slice.X / 4; // Range reduced -> fewer constants
   SliceToSumoProperty SliceToSumo = SliceToSumoMap[XCompact];
 
   SumoPixel Sumo;
   Sumo.X = uint8_t(Slice.X - SliceToSumo.SumoStartOffsetX);
   Sumo.Y = Slice.Y;
   Sumo.Width = SliceToSumo.SumoWidth;
-  Sumo.Sumo = SliceToSumo.SumoId;
+  Sumo.SumoId = SliceToSumo.SumoId;
+  TestEnvAssertMsg(Sumo.IsValid(), "Bad SumoPixel");
   return Sumo;
 }
 
+SlicePixel SlicePixelFromSumoPixel(SumoPixel Sumo, uint32_t SectorIdx,
+                                   uint32_t StripIdx) {
+  TestEnvAssertMsg(Sumo.IsValid(), "Bad SumoPixel");
+  static const uint8_t SumoStartOffsetX[4] = {0, 20, 36, 48};
+  uint32_t SumoIdx = 6 - Sumo.SumoId; // Sumo6=0, Sumo3=3
+  SlicePixel Slice;
+  Slice.X = Sumo.X + SumoStartOffsetX[SumoIdx];
+  Slice.Y = Sumo.Y;
+  Slice.SectorIdx = SectorIdx;
+  Slice.StripIdx = StripIdx;
+  TestEnvAssertMsg(Slice.IsValid(), "Bad SlicePixel");
+  return Slice;
+}
+
 StripPlanePixel StripPlanePixelFromSumoPixel(SumoPixel Sumo) {
+  TestEnvAssertMsg(Sumo.IsValid(), "Bad SumoPixel");
   StripPlanePixel StripPlane;
   uint32_t CassetteCounterIdx = Sumo.Width - Sumo.X - 1;
   StripPlane.CassetteIdx = CassetteCounterIdx / 2;
   StripPlane.CounterIdx = CassetteCounterIdx % 2;
-  StripPlane.WireIdx = SliceHeight - Sumo.Y - 1;
+  StripPlane.WireIdx = DreamGeometry::SliceHeight - Sumo.Y - 1;
+  StripPlane.SumoId = Sumo.SumoId;
+  TestEnvAssertMsg(StripPlane.IsValid(), "Bad StripPlanePixel");
   return StripPlane;
 }
 
-EndCapParams EndCapParamsFromPixelId(uint32_t PixelId) {
+SumoPixel SumoPixelFromStripPlanePixel(StripPlanePixel StripPlane) {
+  TestEnvAssertMsg(StripPlane.IsValid(), "Bad StripPlanePixel");
+  uint32_t CassetteCounterIdx =
+      StripPlane.CassetteIdx * 2 + StripPlane.CounterIdx;
+  SumoPixel Sumo;
+  Sumo.SumoId = StripPlane.SumoId;
+  Sumo.Width = DreamGeometry::SumoWidths[Sumo.SumoId];
+  Sumo.X = Sumo.Width - CassetteCounterIdx - 1;
+  Sumo.Y = DreamGeometry::SliceHeight - StripPlane.WireIdx - 1;
+  TestEnvAssertMsg(Sumo.IsValid(), "Bad SumoPixel");
+  return Sumo;
+}
+
+bool EndCapParamsFromPixelId(uint32_t PixelId, EndCapParams &OutEndCap) {
+  if (!IsPixelIdValid(PixelId)) {
+    return false;
+  }
   SlicePixel Slice = SlicePixelFromPixelId(PixelId);
   SumoPixel Sumo = SumoPixelFromSlicePixel(Slice);
   StripPlanePixel StripPlane = StripPlanePixelFromSumoPixel(Sumo);
 
-  EndCapParams EndCap;
-  EndCap.Sector = Slice.SectorIdx + 1;
-  EndCap.Sumo = Sumo.Sumo;
-  EndCap.Strip = Slice.StripIdx + 1;
-  EndCap.Wire = StripPlane.WireIdx + 1;
-  EndCap.Cassette = StripPlane.CassetteIdx + 1;
-  EndCap.Counter = StripPlane.CounterIdx + 1;
+  OutEndCap.Sector = Slice.SectorIdx + 1;
```

Comment:
Not sure why we need all the + 1's  ?

Reply:
the "Idx" endings are zero based instead of one based. Doing the space conversions got very clunky when it's 1 based.